### ISSUE TITLE
"key-spacing" tab alignment

### ISSUE THREAD

**Issue by brenfwd** [eslint/eslint#12432]
This revisits issues #1727 and #9523.

**What rule do you want to change?**
key-spacing

**Does this change cause the rule to produce more or fewer warnings?**
The proposed change will cause the rule to produce fewer warnings because it will allow the rule to accept tabs as indentation for alignment.

**How will the change be implemented? (New option, new default behavior, etc.)?**
An additional option, `tabs`, would be added, allowing the user to specify whether tabs would be required for alignment indentation, and, if so, how many spaces they should be considered as.

**Please provide some example code that this change will affect:**

```js
// New behavior; allows for tabs with an assumed width of 4 spaces.
/*eslint key-spacing: ["error", { "tab": 4 }]*/
const myObject = {
    property:<tab><tab>value,
    anotherProperty:<tab>value2
};

// Default behavior; disallows tabs. This will raise a warning.
/*eslint key-spacing: ["error", { "tab": "never" }]*/
const myObject = {
    property:<tab><tab>value,
    anotherProperty:<tab>value2
};
```

**What does the rule currently do for this code?**
The rule currently does not treat tabs as valid whitespace for alignment, let alone calculate their width.

**What will the rule do after it's changed?**
After the proposed change, the rule would allow the use of tab characters to more easily align properties.

**Are you willing to submit a pull request to implement this change?**
Yes.

**Comment by kaicataldo** [comment]
I share the same reservations outlined in [this comment](https://github.com/eslint/eslint/issues/1727#issuecomment-71407517). Could you go into more detail how this proposal would handle that?

**Comment by campbellbrendene** [comment]
I would propose that, if the `tab` option is enabled, spaces may not be used to align properties - only tabs. That would allow eslint to internally replace all tabs with *up to* `tabWidth` spaces. [This comment](https://github.com/eslint/eslint/issues/1727#issuecomment-71435145) has some sample code which might work to accomplish this:

> 
>     function convertTabsToSpaces(line, tabWidth) {
>         var chars = line.split(''),
>             len = chars.length,
>             newLine = '',
>             spaces,
>             i,
>             j;
> 
>         for (i = 0; i < len; i++) {
>             if (chars[i] === '\t') {
>                 spaces = 4 - (newLine.length % tabWidth);
>                 for (j = 0; j < spaces; j++) {
>                     newLine += ' ';
>                 }
>             } else {
>                 newLine += chars[i];
>             }
>         }
> 
>         return newLine;
>     }

**Comment by campbellbrendene** [comment]
@kaicataldo thoughts on my previous comment?

**Comment by kaicataldo** [comment]
Sorry I lost track of this - thanks for the ping! This seems reasonable to me (and it looks like the previous issue was accepted). Are you still interested in implementing this if it's accepted?

**Comment by campbellbrendene** [comment]
Sure, I'd gladly work on this. It might be a while before I'll have a PR, since I'm a little swamped right now.

**Comment by fregante** [comment]
If you **expect** tabs to be a certain length, you should use spaces.

I'm all for tabs-for-indentation, but here they're wrong. It would be great **if** editors automatically aligned **one** tab like word processors can (although by manually moving the tab position) but **two** tabs don't make sense. You're just using `tab` as a replacement for 4 spaces.

**Comment by mdjermanovic** [comment]
Closing as this rule has been deprecated per https://github.com/eslint/eslint/issues/17522
